services/api-server/src/simcore_service_api_server/core/health_checker.py

```diff
@@ -0,0 +1,120 @@
+
+
+class ApiServerHealthChecker:
+    def __init__(
+        self,
+        *,
+        registry: CollectorRegistry,
+        log_distributor: LogDistributor,
+        rabbit_client: RabbitMQClient,
+        timeout_seconds: PositiveFloat,
+        allowed_health_check_failures: PositiveInt,
+    ) -> None:
+        self._registry = registry
+        self._log_distributor: LogDistributor = log_distributor
+        self._rabbit_client: RabbitMQClient = rabbit_client
+        self._timeout_seconds = timeout_seconds
+        self._allowed_health_check_failures = allowed_health_check_failures
+
+        self._logstreaming_queues_gauge = Gauge(
+            "log_stream_queue_length",
+            "#Logs in log streaming queue",
+            ["job_id"],
+            namespace=METRICS_NAMESPACE,
+        )
+        self._health_check_qauge = Gauge(
+            "log_stream_health_check",
+            "#Failures of log stream health check",
+            namespace=METRICS_NAMESPACE,
+        )
+        self._health_check_failure_count: NonNegativeInt = 0
+        self._dummy_job_id: JobID = uuid4()
+        self._dummy_queue: asyncio.Queue[JobLog] = asyncio.Queue()
+        self._dummy_message = LoggerRabbitMessage(
+            user_id=UserID(123456789),
+            project_id=self._dummy_job_id,
+            node_id=uuid4(),
+            messages=["Api-server health check message"],
+        )
+        self._background_task: asyncio.Task | None = None
+        _logger.info("Api server health check dummy job_id=%s", f"{self._dummy_job_id}")
+
+    async def setup(self, health_check_task_period_seconds: PositiveFloat):
+        await self._log_distributor.register(
+            job_id=self._dummy_job_id, queue=self._dummy_queue
+        )
+        self._background_task = start_periodic_task(
+            task=self._background_task_method,
+            interval=timedelta(seconds=health_check_task_period_seconds),
+            task_name="api_server_health_check_task",
+        )
+
+    async def teardown(self, timeout_seconds: PositiveFloat):
+        if self._background_task:
+            await stop_periodic_task(self._background_task, timeout=timeout_seconds)
+        await self._log_distributor.deregister(job_id=self._dummy_job_id)
+
+    @property
+    def healthy(self) -> bool:
+        return self._health_check_failure_count <= self._allowed_health_check_failures
+
+    def _increment_health_check_failure_count(self):
+        self._health_check_failure_count += 1
+
+    async def _background_task_method(self):
+        # update prometheus metrics
+        self._health_check_qauge.clear()
+        self._health_check_qauge.set(self._health_check_failure_count)
+        self._logstreaming_queues_gauge.clear()
+        log_queue_sizes = self._log_distributor.get_log_queue_sizes()
+        for job_id, length in log_queue_sizes.items():
+            self._logstreaming_queues_gauge.labels(job_id=job_id).set(length)
+
+        # check health
```

Comment:
This is the "check function" right here. If it the check fails, this information propagates to the orchestrator that can eventually kill the app. 

- Does restarting app actually solves the issue that you are checking?
- Check carefuly that the app closes gracefully (e.g. all resources are deallocated etc)

Reply:
We experienced in the past that the api-server stops consuming logs from rabbitmq and that in this case restarting helps. This check will detect that, so I believe it makes sense. In terms of closing gracefully, I guess the only real issue which can happen is that the background tasks don't cancel. But in that case I handle the raised exceptions.